### PR TITLE
feat(plugin-jwt) rs512

### PR DESCRIPTION
**Pull request by sarraz1**
Add support of RS512 algorithm for signing tokens in jwt plugin.
This add-on is based on the already-used module "crypto".

This pull request contains all the updates submitted by he review of the PR   https://github.com/Mashape/kong/pull/2642 (from next branch).  And this previous one have to be cancelled.

### CODE REVIEW

File: kong/plugins/jwt/daos.lua [hunk 1]
```diff
@@ -12,7 +12,7 @@ local SCHEMA = {
     key = {type = "string", unique = true, default = utils.random_string},
     secret = {type = "string", unique = true, default = utils.random_string},
     rsa_public_key = {type = "string"},
-    algorithm = {type = "string", enum = {"HS256", "RS256", "ES256"}, default = 'HS256'}
+    algorithm = {type = "string", enum = {"HS256", "RS256", "ES256","RS512"}, default = 'HS256'}
```
Comment (thibaultcha) [inline kong/plugins/jwt/daos.lua]: Could we move RS512 next to RS256? Thanks!

File: spec/03-plugins/17-jwt/01-jwt_parser_spec.lua [hunk 2]
```diff
@@ -76,7 +76,29 @@ describe("Plugin: jwt (parser)", function()
...
+        alg = "RS512",
+      })
+      assert(jwt_parser:new(token))
+    end)
+    it("accepts a valid RS512 encoding request with missing TYP", function()
+      local token = jwt_parser.encode({sub = "1234"}, "secret", nil, {alg = "RS512"})
+      assert(jwt_parser:new(token))
+    end)
+
+
```
Comment (thibaultcha) [inline spec/03-plugins/17-jwt/01-jwt_parser_spec.lua]: It seems like this file got many undesired blank lines. Could we remove them?

File: spec/03-plugins/17-jwt/01-jwt_parser_spec.lua [hunk 3]
```diff
@@ -76,7 +76,29 @@ describe("Plugin: jwt (parser)", function()
...
   end)
+
```
Comment (thibaultcha) [inline spec/03-plugins/17-jwt/01-jwt_parser_spec.lua]: here too for example

File: spec/03-plugins/17-jwt/01-jwt_parser_spec.lua [hunk 4]
```diff
@@ -90,6 +112,13 @@ describe("Plugin: jwt (parser)", function()
       assert.True(jwt:verify_signature(fixtures.rs256_public_key))
       assert.False(jwt:verify_signature(fixtures.rs256_public_key:gsub('QAB', 'zzz')))
     end)
+    it("using RS512", function()
+      local token = jwt_parser.encode({sub = "foo"}, fixtures.rs512_private_key, 'RS512')
+      local jwt = assert(jwt_parser:new(token))
+      assert.True(jwt:verify_signature(fixtures.rs512_public_key))
+      assert.False(jwt:verify_signature(fixtures.rs512_public_key:gsub('AE=', 'zzz')))
+    end)
+
```
Comment (thibaultcha) [inline spec/03-plugins/17-jwt/01-jwt_parser_spec.lua]: ditto

File: spec/03-plugins/17-jwt/03-access_spec.lua [hunk 5]
```diff
@@ -232,13 +238,13 @@ describe("Plugin: jwt (access)", function()
     it("verifies JWT", function()
       PAYLOAD.iss = rsa_jwt_secret_1.key
       local jwt = jwt_encoder.encode(PAYLOAD, fixtures.rs256_private_key, 'RS256')
-      local authorization = "Bearer "..jwt
+      local authorization = "Bearer " .. jwt
```
Comment (thibaultcha) [inline spec/03-plugins/17-jwt/03-access_spec.lua]: It doesn't seem like this change is related to the scope of this PR?

File: spec/03-plugins/17-jwt/03-access_spec.lua [hunk 6]
```diff
@@ -232,13 +238,13 @@ describe("Plugin: jwt (access)", function()
...
       local res = assert(proxy_client:send {
         method = "GET",
         path = "/request",
         headers = {
           ["Authorization"] = authorization,
-          ["Host"] = "jwt.com"
+          ["Host"] = "jwt.com",
```
Comment (thibaultcha) [inline spec/03-plugins/17-jwt/03-access_spec.lua]: ditto

File: spec/03-plugins/17-jwt/03-access_spec.lua [hunk 7]
```diff
@@ -263,6 +269,41 @@ describe("Plugin: jwt (access)", function()
     end)
   end)
 
+describe("RS512", function()
+    it("verifies JWT", function()
+      PAYLOAD.iss = rsa_jwt_secret_3.key
+      local jwt = jwt_encoder.encode(PAYLOAD, fixtures.rs512_private_key, 'RS512')
+      local authorization = "Bearer "..jwt
```
Comment (thibaultcha) [inline spec/03-plugins/17-jwt/03-access_spec.lua]: However, this line is missing spaces around the concatenation operator ;)

File: spec/03-plugins/17-jwt/fixtures.lua [hunk 8]
```diff
@@ -39,6 +39,46 @@ FLiGOm5uTMEk8S4txs2efueg1XyymilCKzzuXlJvrvPA4u6HI7qNvuvkvUjQmwBH
 gwIDAQAB
 -----END PUBLIC KEY-----
 ]],
+    rs512_private_key = [[
```
Comment (thibaultcha) [inline spec/03-plugins/17-jwt/fixtures.lua]: The spaces at the beginning of this line are undesired; other keys are not indented like so.

File: spec/03-plugins/17-jwt/03-access_spec.lua (hunk 5 above)
Comment (thibaultcha) [inline spec/03-plugins/17-jwt/03-access_spec.lua]: (style updates of legacy - unrelated - code should not be part of such a feature PR, is what I mean)

File: kong/plugins/jwt/daos.lua [hunk 9]
```diff
@@ -21,6 +21,12 @@ local SCHEMA = {
     if plugin_t.algorithm == "RS256" and crypto.pkey.from_pem(plugin_t.rsa_public_key) == nil then
       return false, Errors.schema "'rsa_public_key' format is invalid"
     end
+    if plugin_t.algorithm == "RS512" and plugin_t.rsa_public_key == nil then
+       return false, Errors.schema "no mandatory 'rsa_public_key'"
```
Comment (p0pr0ck5) [inline kong/plugins/jwt/daos.lua]: looks like we're indenting too far here? indents should be two spaces.

Comment (thibaultcha) [inline kong/plugins/jwt/daos.lua]: Good catch! @sarraz1 Any updates here? Those style issues are all that's left for us to merge it, if you address them we won't hesitate to push the button :)

Comment (sarraz1) [inline kong/plugins/jwt/daos.lua]: Hello, the changes requested were committed.  Regards ;-)

**Comment by thibaultcha** [comment]
Manually merged to master with a few changes in the tests, code style and commit message. Thank you!

**Pull request closed without merging**